python/arcticdb/diagnostics_engine/version_list_diagnosis.py

```diff
@@ -0,0 +1,342 @@
+
+        # Add edge to next node
+        if i < len(parsed_keys) - 1:
+            next_node_name = f"{parsed_keys[i + 1].version_id}_{i + 1}"
+            G.add_edge(node_name, next_node_name)
+
+    plt.figure(figsize=(12, 8))
+    pos = zigzag_layout(G)
+    nx.draw(G, pos, labels=node_labels, node_size=1200, arrows=True, node_color=node_colors)
+    add_graph_description(compacted_versions, exists_in_version_list, is_tombstoned)
+
+
+def parse_versions(versions):
+    """
+    Parse versions into a dictionary for easy access.
+    """
+    parsed_versions = {v["version"]: v for v in versions}
+    return parsed_versions
+
+
+def parse_keys(keys):
+    """
+    Parse and sort keys based on creation timestamp.
+    """
+    parsed_keys = sorted(keys, key=lambda k: k.creation_ts, reverse=True)
+    return parsed_keys
+
+
+def get_target_version(parsed_versions, parsed_keys, as_of):
+    """
+    Get the target version based on the 'as_of', or the latest version if 'as_of' is None.
+    """
+    if as_of is not None:
+        return as_of
+
+    alive_versions = parsed_versions.keys()
+    for key in parsed_keys:
+        if key in alive_versions and parsed_versions[key.version_id]["deleted"] == False:
+            return key.version_id
+
+
+def traverse_version_list(parsed_keys, target):
+    """
+    Traverse the version list to find the target version and the number of steps required to reach it.
+    """
+    steps = 0
+    is_present = False
+    for i, key in enumerate(parsed_keys):
+        steps += 1
+        if key.version_id == target:
+            is_present = True
+            break
+
+    return steps - 1, is_present
+
+
+def get_compacted_versions(parsed_versions, last_version_in_list):
+    """
+    Get all the compacted versions after the last node from the version list.
+    """
+    compacted_versions = []
+    for version in parsed_versions.keys():
+        if parsed_versions[version]["deleted"] == False and version <= last_version_in_list:
+            compacted_versions.append(version)
+    return compacted_versions
+
+
+def determine_where_version_exists(present_in_version_list, parsed_versions, target):
+    """
+    Determine the existence of the target version in various locations: version list, compacted versions, snapshots.
+    """
+    exist_in_version_list = False
+    exists_as_compacted_version = False
+    exists_in_snaphots = False
+    is_tombstoned = False
+
+    if target in parsed_versions.keys():
+        # if it exists either in the version list or compacted, we don't about about if also exists in snapshots
+        if parsed_versions[target]["deleted"] == False:
+            if present_in_version_list:
+                exist_in_version_list = True
+            else:
+                exists_as_compacted_version = True
+        else:
+            if parsed_versions[target]["snapshots"]:
+                exists_in_snaphots = True
+    elif present_in_version_list:
+        is_tombstoned = True
```

Comment:
A version can be deleted but also present in snapshots

Reply:
Yes, that's definitely a bug. I had been considering this scenario all the time, but it wasn't correctly reflected in the logic. Good catch!"